pkg/builds/task.go

```diff
@@ -41,17 +41,18 @@ type CancelableTaskOption func(context.Context, *tknv1beta1.TaskRun) (context.Ca
 
 // RunTask executes the provided TaskRun with the provided options applied, and returns
 // the final TaskRun state (or error) upon completion.
-func RunTask(ctx context.Context, tr *tknv1beta1.TaskRun, opt *options.LogOptions, opts ...CancelableTaskOption) (*tknv1beta1.TaskRun, error) {
-	// TODO(mattmoor): expose masterURL and kubeconfig flags.
-	cfg, err := GetConfig("", "")
-	if err != nil {
-		return nil, err
-	}
-	client, err := tektonclientset.NewForConfig(cfg)
-	if err != nil {
-		return nil, err
+func RunTask(ctx context.Context, client tektonclientset.Interface, tr *tknv1beta1.TaskRun, opt *options.LogOptions, opts ...CancelableTaskOption) (*tknv1beta1.TaskRun, error) {
+	if client == nil {
+		// TODO(mattmoor): expose masterURL and kubeconfig flags.
+		cfg, err := GetConfig("", "")
+		if err != nil {
+			return nil, err
+		}
+		client, err = tektonclientset.NewForConfig(cfg)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
So I was sort of leaning towards using the context injection stuff to set up clients on `ctx` so if you want a tekton client could can just use: https://github.com/tektoncd/pipeline/blob/77dc827e42dfe7745eee5484a990f4fa99fcb33d/pkg/reconciler/taskrun/controller.go#L51 or a kubernetes client: https://github.com/tektoncd/pipeline/blob/77dc827e42dfe7745eee5484a990f4fa99fcb33d/pkg/reconciler/taskrun/controller.go#L50

Reply:
We'd want to set up the context like so: https://github.com/knative/pkg/blob/af53027cbd910ea11d24defd16bbe660b33c21d3/injection/sharedmain/main.go#L176

You can set up fake clients with: https://github.com/tektoncd/pipeline/blob/77dc827e42dfe7745eee5484a990f4fa99fcb33d/vendor/knative.dev/pkg/reconciler/testing/context.go#L43

Reply:
ooh thats really nice, I'd not noticed that before thanks! I'll take a stab at it...